### ISSUE TITLE
Wild and Crazy URLs

### ISSUE THREAD

**Issue by robertdfrench** [LLNL/spack#1757]
Hello friends!

Spack has trouble recognizing this url as a tar.gz file: 
`http://www.ftp.cpc.ncep.noaa.gov/wd51we/wgrib2/wgrib2.tgz.v2.0.5`

It says it "cannot deduce" the filetype. That phrase led me here: [lib/spack/spack/url.py:164](https://github.com/LLNL/spack/blob/f59653ac2c9b20ec5954d90fda019c7652644ac9/lib/spack/spack/url.py#L164)

It looks like `url.download_file_extension` enforces some sanity around the extension. My question is, what is my best bet for dealing with packages that have uncivilized extensions like ".tgz.v2.0.5"? Is it appropriate to look for `tgz` *near* the end of the path? That's sortof a weird criterion...

Anyhow, feedback is welcome. Here is a reproducer package:

```python
from spack import *


class WgribCrazyUrl(Package):
    """WGRIB is a program to manipulate, inventory and decode GRIB files"""

    homepage = "http://www.cpc.ncep.noaa.gov/products/wesley/wgrib.html"
    url      = "http://www.ftp.cpc.ncep.noaa.gov/wd51we/wgrib2/wgrib2.tgz.v2.0.5"

    version('2.0.5', '84029e969b5b37e1ba791d0572895133')

    parallel = False

    def install(self, spec, prefix):
        gmake("FC=fcc CC=cc")

        install_tree('bin', prefix.bin)
        install_tree('lib', prefix.lib)
        install_tree('include', prefix.include)
        install_tree('man/man1', prefix.share_man1)

        install('wgrib2/wgrib2', prefix.bin)
```

**Comment by scheibelp** [comment]
Made some tweaks to get this to stage in #1758 

It may take a day or two for me to tweak things (then some additional time for it to be merged into develop)

**Comment by scheibelp** [comment]
At this point the strategy is implemented as well as it could be in #1758 but I want to leave it as a WIP to consider other approaches that may simplify the fetching logic while satisfying this use case (that will take several more days as it involves more fiddling with Spack internals). In the meantime I hope it is not too obnoxious to pull that PR for building wgrib (or continuing to use your own solution if you have found one) - an example is now included in the PR description

**Comment by tgamblin** [comment]
@robertdfrench: can you try this with #1758 merged?
